### ISSUE TITLE
Segmentation fault (core dumped) when indexing hg19, hg38

### ISSUE THREAD

**Issue by noblerabbit** [bwa-mem2/bwa-mem2#19]
Hi,

I tried with pre-compiled and compiled (hg19 and hg38). The same output:
```
bwa-mem2/bwa-mem2 index Homo_sapiens_assembly19.fasta                                                                                                                 
[bwa_index] Pack FASTA... 13.35 sec
init ticks = 204490443602
ref seq len = 6203953124
binary seq ticks = 117021996235
build index ticks = 6870572410362
ref_seq_len = 6203953124
count = 0, 1811627992, 3101976562, 4392325132, 6203953124
Segmentation fault (core dumped)
```

Is this being worked on? Where can I get the index files for the reference genome suitable for bwa-mem2.

Thanks.

**Comment by yuk12** [comment]
Please try it with the latest commit.

**Comment by noblerabbit** [comment]
It was already the latest, but I tried again. Still the same outcome:

```
$ bwa-mem2/bwa-mem2 index Homo_sapiens_assembly19.fasta
[bwa_index] Pack FASTA... 11.09 sec
init ticks = 206750171834
ref seq len = 6203953124
binary seq ticks = 117289660895
build index ticks = 6871280898905
ref_seq_len = 6203953124
count = 0, 1811627992, 3101976562, 4392325132, 6203953124
Segmentation fault (core dumped)
```

`/bwa-mem2$ git pull
Already up to date.`

`commit 2c91e6570a09a672554542f097dc404ed9ca1d4e (HEAD -> master, origin/master, origin/HEAD)
Author: Vasimuddin <wasim.mzr@gmail.com>
Date:   Mon Sep 16 01:24:41 2019 -0700
fixed failed assert 2
`

**Comment by fredjarlier** [comment]
Hi I think it's a matter of memory how much do you have? 
maybe a memory leaks...

f

**Comment by noblerabbit** [comment]
@fredjarlier I have 64 GB (16 core, AMD CPU). I was looking at memory. It never reached even 10%. I have now tried also on a different machine with 32 GB (8 core, Intel i7). I get the same result.

**Comment by fredjarlier** [comment]
do you have the core file?

**Comment by noblerabbit** [comment]
@fredjarlier what do you mean by core file - Homo_sapiens_assembly19.fasta?

If I run the same file on the bwa tool, indexing works...

```
$ bwa/bwa index reference/Homo_sapiens_assembly19.fasta
[bwa_index] Pack FASTA... 23.16 sec
[bwa_index] Construct BWT for the packed sequence...
[BWTIncCreate] textLength=6203953124, availableWord=448532900
[BWTIncConstructFromPacked] 10 iterations done. 99999988 characters processed.
[BWTIncConstructFromPacked] 20 iterations done. 199999988 characters processed.
[BWTIncConstructFromPacked] 30 iterations done. 299999988 characters processed.
```

**Comment by fredjarlier** [comment]
Sorry I mean the core dump you can produce it with gcore or valgrind ... 
like gcore -a -o mydump pid_process.

I was also thinking maybe you have ulimit?

**Comment by mp15** [comment]
@noblerabbit I suspect this is because bwa-mem2 is failing to check for success on it's large mallocs. I assume _mm_malloc is returning a NULL pointer when the alloc fails and thus your segfault. Perhaps for the largest allocs it might be an idea to add a few `if (allocptr == NULL) { perror("Allocation of table x failed"); exit(EXIT_FAILURE);}`

Incidentally what is the advantage of using _mm_malloc now that we have aligned_alloc?

**Comment by mp15** [comment]
@noblerabbit I've just embodied this in a pull request. Try my pull request #20 and if I am right you should get out of memory errors rather than segfaults.

**Comment by tonisolu** [comment]
I just ran into same issue and with that patch applied:
#/work/tools/bwa-mem2/bwa-mem2 index hg38.p12.fa
[bwa_index] Pack FASTA... 9.59 sec
init ticks = 217929621292
ref seq len = 6514694564
binary seq ticks = 124595597266
Allocation of suffix_array failed: Cannot allocate memory

Looking at strace it asks:
mmap(NULL, 52117557248, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = -1 ENOMEM (Cannot allocate memory)
brk(0x56277a75a000)                     = 0x561b5802e000
mmap(NULL, 52117688320, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = -1 ENOMEM (Cannot allocate memory)

So it wants around 52 gigs of more memory.. is there a way to construct this index with less memory?

**Comment by mp15** [comment]
@tonisolu Then the patch works, all the patch is supposed to do is allow you to distinguish a bug in BWA mem and thus a segfault from running out of memory.

Some of the speedups you get from BWA MEM2 are a traded off against memory consumption (they decompressed the suffix index). I would recommend using the original BWA MEM if you want to use less memory, having 52 gigs isn't abnormal for a modern HPC cluster.

**Comment by tonisolu** [comment]
That 52 gigs were on top of around 48 gigs it was using already. Anyway it's all good, I was just testing out the new stuff. So yes that patch works as intended.

**Comment by da-i** [comment]
I've run into the same issue,

`
# get precompiled tool
curl -L https://github.com/bwa-mem2/bwa-mem2/releases/download/v2.0pre1/bwa-mem2-2.0pre1_x64-linux.tar.bz2 | tar jxf -
PATH="$PATH:$(pwd)/bwa-mem2-2.0pre1_x64-linux/"
# download official release from broad
gsutil cp gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta .
bwa-mem2 index Homo_sapiens_assembly38.fasta
`

results in
```[bwa_index] Pack FASTA... 13.91 sec
init ticks = 280061516970
ref seq len = 6434693834
binary seq ticks = 246960337754
Segmentation fault (core dumped)```

**Comment by da-i** [comment]
please see issue #9 as it explains that you need 34N memory to build the index

**Comment by sanchit-misra** [comment]
@mp15 I have applied the pull request #20. Thanks for that. Now, the code reports an error if it was unable to allocate memory and the name of the variable the memory. I have also added reporting of how much memory was requested.
@tonisolu @da-i The code needed 34N GB memory to build the index where N is the size of the reference sequence. So, for N = 3 Billion (for human genome), it would have needed nearly 100 GB memory. The latest commit (<code>6f21de2873a892207c4d315a3a0e59cfffb0b71b</code>) reduces the memory requirement to 28N GB (=84 GB for human genome).
